### PR TITLE
Adjust runtime syscalls to NEO3

### LINKED ISSUE #1028: Adjust runtime syscalls to Neo 3
```
Platform = Register("System.Runtime.Platform", Runtime_Platform, 0_00000250, TriggerType.All, CallFlags.None);
GetTrigger = Register("System.Runtime.GetTrigger", Runtime_GetTrigger, 0_00000250, TriggerType.All, CallFlags.None);
GetTime = Register("System.Runtime.GetTime", Runtime_GetTime, 0_00000250, TriggerType.Application, CallFlags.None);
GetScriptContainer = Register("System.Runtime.GetScriptContainer", Runtime_GetScriptContainer, 0_00000250, TriggerType.All, CallFlags.None);
GetExecutingScriptHash = Register("System.Runtime.GetExecutingScriptHash", Runtime_GetExecutingScriptHash, 0_00000400, TriggerType.All, CallFlags.None);
GetCallingScriptHash = Register("System.Runtime.GetCallingScriptHash", Runtime_GetCallingScriptHash, 0_00000400, TriggerType.All, CallFlags.None);
GetEntryScriptHash = Register("System.Runtime.GetEntryScriptHash", Runtime_GetEntryScriptHash, 0_00000400, TriggerType.All, CallFlags.None);
CheckWitness = Register("System.Runtime.CheckWitness", Runtime_CheckWitness, 0_00030000, TriggerType.All, CallFlags.None);
GetInvocationCounter = Register("System.Runtime.GetInvocationCounter", Runtime_GetInvocationCounter, 0_00000400, TriggerType.All, CallFlags.None);
Log = Register("System.Runtime.Log", Runtime_Log, 0_01000000, TriggerType.All, CallFlags.AllowNotify);
Notify = Register("System.Runtime.Notify", Runtime_Notify, 0_01000000, TriggerType.All, CallFlags.AllowNotify);
GetNotifications = Register("System.Runtime.GetNotifications", Runtime_GetNotifications, 0_00010000, TriggerType.All, CallFlags.None);
GasLeft = Register("System.Runtime.GasLeft", Runtime_GasLeft, 0_00000400, TriggerType.All, CallFlags.None);
```
`Platform`, `GasLeft` and `GetInvocationCounter` need to be implemented, everything else just needs to be adjusted for prices/flags.

### PR DESCRIPTION
**Pull request by fyrchik**
Closes #1028 .

### CODE REVIEW

File: pkg/core/interop/runtime/util.go [hunk 1]
```diff
@@ -2,11 +2,50 @@ package runtime
...
+		}
+	}
+	if len(notifications) > vm.MaxStackSize {
+		return errors.New("notification count is too big")
+	}
+	arr := stackitem.NewArray(make([]stackitem.Item, 0, len(notifications)))
+	for i := range notifications {
+		ev := stackitem.NewArray([]stackitem.Item{
+			stackitem.NewByteArray(notifications[i].ScriptHash.BytesBE()),
+			notifications[i].Item,
```
Comment (roman-khimov) [inline pkg/core/interop/runtime/util.go:45]: Suppose it's an `Array`, now I can change it in the script and affect already emitted notification.

File: pkg/core/interop/runtime/util.go [hunk 2]
```diff
@@ -2,11 +2,50 @@ package runtime
...
+		}
+		notifications = []state.NotificationEvent{}
+		for i := range ic.Notifications {
+			if ic.Notifications[i].ScriptHash.Equals(u) {
+				notifications = append(notifications, ic.Notifications[i])
+			}
+		}
+	}
+	if len(notifications) > vm.MaxStackSize {
+		return errors.New("notification count is too big")
```
Comment (roman-khimov) [inline pkg/core/interop/runtime/util.go:39]: `too many notifications`

**Comment by roman-khimov** [comment]
And btw, all the new syscalls should be available to Go contracts.

File: pkg/core/interop/runtime/util.go (hunk 1 above)
Comment (fyrchik) [reply pkg/core/interop/runtime/util.go:45]: I thought about this, but it seems that the original behavior is the same. In C# implementation no copies are made in both `Notify` and `GetNotifications` syscalls.

File: pkg/core/interop/runtime/util.go (hunk 2 above)
Comment (fyrchik) [reply pkg/core/interop/runtime/util.go]: Fixed.

File: pkg/vm/vm.go [hunk 3]
```diff
@@ -134,6 +134,11 @@ func (v *VM) GasConsumed() util.Fixed8 {
 	return v.gasConsumed
 }
 
+// GasLimit returns vm's gas limit.
+func (v *VM) GasLimit() util.Fixed8 {
+	return v.gasLimit
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/vm/vm.go:139]: Done.

File: pkg/core/interop/runtime/util.go (hunk 1 above)
Comment (roman-khimov) [reply pkg/core/interop/runtime/util.go:45]: Hmm. Can we try some fancy notify-getnotifications-change-them on testnet? I don't think anyone intended these notifications to be updateable.

File: pkg/compiler/codegen.go [hunk 4]
```diff
@@ -586,6 +586,8 @@ func (c *codegen) Visit(node ast.Node) ast.Visitor {
 	case *ast.Ident:
 		if tv := c.typeAndValueOf(n); tv.Value != nil {
 			c.emitLoadConst(tv)
+		} else if n.Name == "nil" {
+			emit.Opcode(c.prog.BinWriter, opcode.PUSHNULL)
```
Comment (roman-khimov) [inline pkg/compiler/codegen.go:590]: It's not about `GetNotifications`.

File: pkg/interop/runtime/runtime.go [hunk 5]
```diff
@@ -59,3 +59,11 @@ func Verification() byte {
 func GasLeft() int64 {
 	return 0
 }
+
+// GetNotifications returns notifications emitted by contract h.
+// 'nil' literal means no filtering. It returns slice consisting of following elements:
+// [  scripthash of notification's contract  ,  emitted item  ].
+// TODO update to return slice of pointers to structs after #1032
```
Comment (roman-khimov) [inline pkg/interop/runtime/runtime.go:66]: Not sure this TODO is useful here, better add a comment to #1032 and tell which syscall is being used here.

File: pkg/core/interop/runtime/util.go [hunk 6]
```diff
@@ -0,0 +1,12 @@
+package runtime
+
+import (
+	"github.com/nspcc-dev/neo-go/pkg/core/interop"
+	"github.com/nspcc-dev/neo-go/pkg/vm"
+)
+
+// GasLeft returns remaining amount of GAS.
```
Comment (roman-khimov) [inline pkg/core/interop/runtime/util.go:12]: Missing syscall in the description.

File: pkg/core/interop/runtime/util.go [hunk 7]
```diff
@@ -49,3 +49,13 @@ func GetNotifications(ic *interop.Context, v *vm.VM) error {
 	v.Estack().PushVal(arr)
 	return nil
 }
+
+// GetInvocationCounter returns how many times current contract was invoked during current tx execution.
```
Comment (roman-khimov) [inline pkg/core/interop/runtime/util.go:53]: Syscall.

Comment (roman-khimov) [reply pkg/core/interop/runtime/util.go:53]: Not exactly here, but anyway.

**Comment by roman-khimov** [comment]
BTW, we're also missing `pkg/interop` wrappers for syscalls added in #1054.

File: pkg/core/interop/runtime/util.go (hunk 1 above)
Comment (fyrchik) [reply pkg/core/interop/runtime/util.go:45]: Try this script `57040012111013c07068115511c04195016f61210b115541274335f121716910ce11ce10ce7200286a1151d06811ce6812ce9e40` .
It is for the following contract:
```golang
package enginecontract

import (
	"github.com/nspcc-dev/neo-go/pkg/interop/runtime"
)

// Main is that famous Main() function, you know.
func Main() int {
	a := []int{0, 1, 2}
	runtime.Notify(a)

	notifications := runtime.GetNotifications(nil)
	notif := notifications[0][1].([]interface{})[0].([]int)
	notif[1] = 40
	return a[1] + a[2]
}
```

Got following response:
```json
{
    "jsonrpc": "2.0",
    "id": 1,
    "result": {
        "script": "57040012111013c07068115511c04195016f61210b115541274335f121716910ce11ce10ce7200286a1151d06811ce6812ce9e40",
        "state": "FAULT",
        "gas_consumed": "2646850",
        "stack": [
            {
                "type": "Integer",
                "value": "42"
            }
        ]
    }
}
```

Maybe it is a bug, but we should be compatible with `preview2`.

Comment (roman-khimov) [reply pkg/core/interop/runtime/util.go:45]: > Got following response

Argh, it doesn't show notifications for some reason (looks like this field was removed).

> we should be compatible with preview2.

No doubt about that, but I think it's worth submitting a bug for neo.

Comment (fyrchik) [reply pkg/core/interop/runtime/util.go:45]: https://github.com/neo-project/neo/issues/1707

File: pkg/interop/runtime/runtime.go (hunk 5 above)
Comment (fyrchik) [reply pkg/interop/runtime/runtime.go]: Fixed.

File: pkg/compiler/codegen.go (hunk 4 above)
Comment (fyrchik) [reply pkg/compiler/codegen.go:590]: Fixed.

File: pkg/core/interop/runtime/util.go (hunk 7 above)
Comment (fyrchik) [reply pkg/core/interop/runtime/util.go:53]: Fixed.

**Pull request merged**